glue/clients/tests/test_scatter_client.py

```diff
@@ -699,3 +707,124 @@ def test_log_sticky(self):
+        self.client.xatt = self.ids[0]
+        assert self.client.axes.get_xlabel() == self.ids[0].label
+        self.client.yatt = self.ids[1]
+        assert self.client.axes.get_ylabel() == self.ids[1].label
+
+    def test_range_rois_preserved(self):
+        data = self.add_data_and_attributes()
+        assert self.client.xatt is not self.client.yatt
+
+        roi = core.roi.XRangeROI()
+        roi.set_range(732678.0, 733408.0)
+        self.client.apply_roi(roi)
+        assert isinstance(data.edit_subset.subset_state,
+                          core.subset.RangeSubsetState)
+        assert data.edit_subset.subset_state.att == self.client.xatt
+
+        roi = core.roi.YRangeROI()
+        roi.set_range(2, 5)
+        self.client.apply_roi(roi)
+        assert data.edit_subset.subset_state.att == self.client.yatt
+
+    def test_apply_roi(self):
+        data = self.add_data_and_attributes()
+        roi = core.roi.RectangularROI()
+        roi.update_limits(*self.roi_limits)
+        x, y = self.roi_points
+        self.client.apply_roi(roi)
+        assert self.layer_data_correct(data.edit_subset, x, y)
+
+    def test_apply_roi_adds_on_empty(self):
+        data = self.add_data_and_attributes()
+        data._subsets = []
+        data.edit_subset = None
+        roi = core.roi.RectangularROI()
+        roi.update_limits(*self.roi_limits)
+        x, y = self.roi_points
+        self.client.apply_roi(roi)
+        assert data.edit_subset is not None
+
+    def test_apply_roi_applies_to_all_editable_subsets(self):
+        d1 = self.add_data_and_attributes()
+        d2 = self.add_data()
+        state1 = d1.edit_subset.subset_state
+        state2 = d2.edit_subset.subset_state
+        roi = core.roi.RectangularROI()
+        roi.update_limits(*self.roi_limits)
+        x, y = self.roi_points
+        self.client.apply_roi(roi)
+        assert d1.edit_subset.subset_state is not state1
+        assert d1.edit_subset.subset_state is not state2
+
+    def test_apply_roi_doesnt_add_if_any_selection(self):
+        d1 = self.add_data_and_attributes()
+        d2 = self.add_data()
+        d1.edit_subset = None
+        d2.edit_subset = d2.new_subset()
+        ct = len(d1.subsets)
+        roi = core.roi.RectangularROI()
+        roi.update_limits(*self.roi_limits)
+        x, y = self.roi_points
+        self.client.apply_roi(roi)
+        assert len(d1.subsets) == ct
+
+    def test_check_formatter(self):
+        self.add_data()
+        self.client.xatt = self.ids[0]
+        self.client.yatt = self.ids[0]
+        axes = self.client.axes
+        xformat = axes.xaxis.get_major_formatter()
+        assert isinstance(xformat, AutoDateFormatter)
+
+    def test_check_locator(self):
+        self.add_data()
+        self.client.xatt = self.ids[0]
+        self.client.yatt = self.ids[0]
+        axes = self.client.axes
+        xformat = axes.xaxis.get_major_locator()
+        assert isinstance(xformat, AutoDateLocator)
```

Comment:
Can you add a test that applies a PolygonalROI with one normal component and one datetime component, and makes sure the subset.to_mask() method works? I get an error when I make such a scatter plot and trace a shape:
```
  File "/Users/beaumont/glue/glue/core/subset.py", line 436, in to_mask
    result = self.roi.contains(x, y)
  File "/Users/beaumont/glue/glue/core/roi.py", line 508, in contains
    xypts = np.column_stack((x.flat, y.flat))
  File "/Users/beaumont/anaconda/lib/python2.7/site-packages/numpy/lib/shape_base.py", line 317, in column_stack
    return _nx.concatenate(arrays, 1)
TypeError: invalid type promotion
```